Manual instrumentation documentation (Orbit.h)

pierricgimmig opened the pull request.

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
 #define ORBIT_SCOPE_WITH_COLOR(name, col) orbit_api::Scope ORBIT_VAR(name, col)
 
-// ORBIT_START/ORBIT_STOP: profile time span on a single thread.
+// ORBIT_START/ORBIT_STOP: Profile sections inside a scope.
```
Similarly here, also mention ORBIT_{START, STOP}_WITH_COLOR here?

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
+//
+// Parameters:
+// name: [const char*] Label to be displayed on current time slice (kMaxEventStringSize characters).
+// col: [orbit::Color] User defined color for the current time slice.
```
Maybe point out that predefined colors can be found below? 

"(see below for color definitions)"

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
 #if ORBIT_API_ENABLED
 
-// ORBIT_SCOPE: profile current scope.
+// ORBIT_SCOPE: Profile current scope.
```
Alternatively, mention further above that many of the macros have a _WITH_COLOR variant that allows specifying the timeslice color.

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
+// ORBIT_START_ASYNC/ORBIT_STOP_ASYNC: Profile time spans across scopes or threads.
+//
+// Overview:
+// Async time spans can be started in one scope and stopped on another. They will be displayed
```
"stopped in another" (scope)

ronaldfw replied:
Seems this was not addressed?

ronaldfw replied:
Looked at the wrong version, sorry.

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
+// Note:
+// We limit the maximum number of characters of the "name" parameter to "kMaxEventStringSize". This
+// limitation will be lifted as we roll out a new dynamic instrumentation implementation on Linux.
+// It is possible however to add per-time slice additional information using the ASYNC_STRING macro.
```
Should we point out performance considerations when using the ASYNC_STRING macro?

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
+//
+// Parameters:
+// name: [const char*] Name of the *track* that will display the async events in Orbit.
+// id: [uint64_t] A user-provided unique id for the time slice.
```
I think it could help understanding here to add how the unique id is used: "This unique id is used to match start and stop calls."

IMO, that clearly emphasizes that it *has* to be unique and strange stuff will happen if it is not.

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
 #define ORBIT_START_ASYNC_WITH_COLOR(name, id, col) orbit_api::StartAsync(name, id, col)
 #define ORBIT_STOP_ASYNC(id) orbit_api::StopAsync(id)
+
+// ORBIT_ASYNC_STRING: Provide additional information for an async time span.
```
The "additional information" sounds like more than it is (a string) and would initially confuse me as a user. Maybe just say: "Provide an additional string for an async time span."?

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
+// Thread 1: ORBIT_START_ASYNC("File IO", unique_64_bit_id);
+// Thread 1 or 2: ORBIT_ASYNC_STRING(unique_64_bit_id, "My very long file path");
+// Thread 1 or 2: ORBIT_STOP_ASYNC(unique_64_bit_id);
+//
```
IMO, it would help to state what will be displayed on the timeslice in this example.

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
+// ORBIT_[type]: Graph variables.
+//
+// Overview:
+// Send values to be plotted in a track uniquely identified by "name".
```
Mention that these are plotted over time? It says this below, but it's easier to understand if you state it here already.

ronaldfw reviewed (approved):
Thanks for this, @pierricgimmig! I have mostly minor suggestions, feel free to address what you consider important, and feel free to merge as you see fit!

pierricgimmig commented:
Hi Ronald, thanks for the comments. I addressed all of them and added a bit more information on the experimental nature of the feature and on the general performance overhead which I'd like to pass by you to double check. Thanks again!

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,203 @@
+// Thread 1: ORBIT_START_ASYNC("File IO", unique_64_bit_id);  // File IO request site.
+// Thread 1 or 2: ORBIT_ASYNC_STRING(unique_64_bit_id, "My very long file path");
+// Thread 1 or 2: ORBIT_STOP_ASYNC(unique_64_bit_id);  // File IO result site.
+// Result: Multiple time slices labeled with the results of "io_request->GetFileName()" will appear
```
I don't understand the relation of "io_request->GetFileName()" to the above example. Am I missing something?

ronaldfw reviewed (approved):
LGTM, just one question left.

ronaldfw commented on Orbit.h:
```diff
@@ -9,45 +9,173 @@
+//
+// Note:
+// We limit the maximum number of characters of the "name" parameter to "kMaxEventStringSize". This
+// limitation will be lifted as we roll out a new dynamic instrumentation implementation on Linux.
```
Thoughts on this?

ronaldfw replied:
Sorry, was looking at the wrong version, this was addressed, thanks.

ronaldfw commented:
Merged as we need this for 1.53.
